Polis Integration

cronos-polis opened the pull request:
Experimental Polis Integration.
We launch btcpayserver only using Polis. Will setup for testing.

NicolasDorier commented:
note that as DOGECOIN, merchant won't be able to use it because there is no source rate.

cronos-polis commented:
What you mean with source rate?

NicolasDorier commented:
If your merchant create an invoice of 10 USD, BTCpay has not way to convert in POLIS because  https://apiv2.bitcoinaverage.com/indices/global/ticker/short does not list the ticker.

cronos-polis commented:
Ok. We can wait the merge then. Will try to add an API for coinmarketcap.

lepipele commented:
@cronos-polis I've talked with @NicolasDorier about refactoring rates fetching logic. I'll hit you up on Slack with details... I do think it makes sense to offer Coinmaketcap API as alternative, but we want to do it in a way that makes it easy to plug in. Currently our merchants are used to Coinaverage which unfortunately doesn't have good niche support.

NicolasDorier commented:
Actually I made https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Services/Rates/BTCPayRateProviderFactory.cs#L68

For being able later to use direct exchange API to access the rate instead of CoinAverage (or by using coinaverage as a fallback).

lepipele commented:
Damn man, you already refactored it completely. I'll hit you up once I'm done with #122 wanna understand few design decisions on that.

NicolasDorier commented:
Yes, it is still not perfect though, and also there is no way to select different exchange depending on the currency. Still work to do on this, but this is good first step.

cronos-polis commented:
Wow, you guys are fast. @lepipele I'll contact you on slack.

NicolasDorier commented:
Now the rate fetching is scriptable. You need to tell use though how rate should be calculated for POLIS in BTCPay out of the box.

NicolasDorier commented:
See https://twitter.com/BtcpayServer/status/992032506944700418

NicolasDorier commented:
mmh I need to refactor. Your exchange will be replaced when the list will be refreshed (every 5 min after boot)

NicolasDorier commented:
Ok done, add you exchange like I did for CoinAverage

https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR98

and 

https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR78

cronos-polis commented:
@NicolasDorier 
We added the exchange to BTCPayRateProviderFactory.cs
Seems that the cryptopia exchange is already supported to ExchangeSharp.
We added this line:
```
            DirectProviders.Add("cryptopia", new ExchangeSharpRateProvider("cryptopia", new ExchangeCryptopiaAPI(), false));
````

https://github.com/jjxtra/ExchangeSharp/blob/master/ExchangeSharp/API/Exchanges/ExchangeCryptopiaAPI.cs

NicolasDorier commented on BTCPayServer/Services/Rates/CoinAverageSettings.cs:
```diff
@@ -121,6 +121,7 @@ public CoinAverageSettings()
                 (DisplayName: "Quoine", Name: "quoine"),
                 (DisplayName: "BTC Markets", Name: "btcmarkets"),
                 (DisplayName: "Bitso", Name: "bitso"),
+                (DisplayName: "Cryptopia", Name: "cryptopia"),
```
remove this line

NicolasDorier commented:
Remove the line I pointed out and then add it here https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR98

cronos-polis commented:
Done

NicolasDorier commented:
still need to add the line where I said.

NicolasDorier commented on BTCPayServer/BTCPayNetworkProvider.Polis.cs:
```diff
@@ -0,0 +1,34 @@
+                },
+                CryptoImagePath = "imlegacy/polis.png",
+                DefaultSettings = BTCPayDefaultSettings.GetDefaultSettings(NetworkType),
+                CoinType = NetworkType == NetworkType.Mainnet ? new KeyPath("3'") : new KeyPath("1'")
```
incorrect KeyPath, ask to add yours on https://github.com/satoshilabs/slips/blob/master/slip-0044.md

NicolasDorier commented:
I added cryptopia in my own commit on https://github.com/btcpayserver/btcpayserver/commit/449738414b1188691540c5fa0d7149dcac4cb1af, please rebase your changes

cronos-polis commented:
@NicolasDorier Done we are waiting the PR from Keypath

NicolasDorier commented:
very good !

cronos-polis commented:
Hello @NicolasDorier Any news on this?
Is there anything we need to do?

NicolasDorier commented:
sorry I forgot about it

NicolasDorier commented:
POLIS support removed because of cryptopedia being hacked 884acdd and breaking btcpay tests

eabz commented:
@NicolasDorier We will look towards the reintegration. 
Thanks.

NicolasDorier commented:
@eabz I will not do it myself, so please make a PR to fix it.
